deeptime/src/include/deeptime/markov/msm/tram/tram.h

```diff
@@ -20,9 +20,43 @@ template<typename dtype>
 constexpr static dtype logPrior() { return -std::numeric_limits<dtype>::infinity(); }
 }
 
+// Compute the sample likelihood from the traminput. This method only gets called by the TRAM struct, not by pybind.
+template<typename dtype>
+static const dtype computeSampleLikelihood(const TRAMInput<dtype> &input,
+                                           const np_array_nfc<dtype> &modifiedStateCountsLog) {
+    auto nThermStates = input.nThermStates();
+    const auto &cumNSamples = input.cumNSamples();
+    std::vector<dtype> sampleWeights(input.nSamples());
+
+    std::vector<ArrayBuffer<BiasMatrix<dtype>, 2>> biasMatrixBuf{begin(input.biasMatrices()),
+                                                                 end(input.biasMatrices())};
+    auto biasMatrixPtr = biasMatrixBuf.data();
+
+    ArrayBuffer<np_array_nfc<dtype>, 2> modifiedStateCountsLogBuf{modifiedStateCountsLog};
+    auto modifiedStateCountsLogPtr = &modifiedStateCountsLogBuf;
+
+    auto inputPtr = &input;
+    #pragma omp parallel for default(none) firstprivate(nThermStates, inputPtr, biasMatrixPtr, sampleWeights, \
+                                                        modifiedStateCountsLogPtr, cumNSamples)
+    for (auto i = 0; i < inputPtr->nMarkovStates(); ++i) {
+        std::vector<dtype> scratch(nThermStates);
+        for (auto x = 0; x < inputPtr->nSamples(i); ++x) {
+            int o = 0;
+            for (StateIndex l = 0; l < nThermStates; ++l) {
+                if ((*modifiedStateCountsLogPtr)(l, i) > -std::numeric_limits<dtype>::infinity()) {
+                    scratch[o++] = (*modifiedStateCountsLogPtr)(l, i) - biasMatrixPtr[i](x, l);
+                }
+            }
+            auto log_divisor = numeric::kahan::logsumexp_sort_kahan_inplace(scratch.begin(), o);
+            sampleWeights[cumNSamples[i] + x] = -log_divisor;
+        }
+    }
```

Comment:
Not sure about this but it might be that the sample weights vector is copied into each thread scope. Maybe openmp treats it as reference, I don't know. Can you check that?

Reply:
also would you mind calling it `logDivisor`?

Reply:
the tests ran so it seems that all sample weigths are computed properly. But I made it shared just in case.